File: android/app/src/main/java/com/openland/spacex/store/readFromStore.kt
package com.openland.spacex.store

import com.openland.spacex.model.*
import org.json.JSONArray
import org.json.JSONObject

private fun selectorKey(value: InputValue, arguments: JSONObject): String? {
    if (value is InputValue.Int) {
        return value.value.toString()
    } else if (value is InputValue.Reference) {
        if (arguments.has(value.name)) {
            val ex = arguments.get(value.name)
            if (ex == null) {
                return "null"
            } else if (ex is String) {
                return "\"" + ex + "\""
            } else if (ex is Number) {
                return ex.toString()
            } else if (ex is Boolean) {
                return ex.toString()
            } else if (ex is JSONArray) {
                val res = mutableListOf<String>()

                return "[]"
            } else {
                error("Unknown input value")
            }
        } else {
            return null
        }
    } else {
        error("Unknown input value")
    }
}

private fun selectorKey(name: String, fieldArguments: Map<String, InputValue>, arguments: JSONObject): String {
    if (fieldArguments.isEmpty()) {
        return name
    }
    val mapped = fieldArguments.mapValues { selectorKey(it.value, arguments) }
    val sortedKeys = mapped.keys.sortedBy { it }
    val converted = mutableListOf<String>()
    for (k in sortedKeys) {
        if (mapped[k] != null) {
            converted.add(k + ":" + mapped[k])
        }
    }
    return name + "(" + converted.joinToString(",") + ")"
}

private fun readValue(value: RecordValue, type: OutputType, store: RecordStore, arguments: JSONObject): Pair<Boolean, Any?> {
    if (type is OutputType.Scalar) {
        if (value == RecordValue.Null) {
            return true to null
        } else if (type.name == "String") {
            return if (value is RecordValue.String) {
                true to value.value
            } else {
                false to null
            }
        } else if (type.name == "Int") {
            return if (value is RecordValue.Number) {
                true to value.value
            } else {
                false to null
            }
        } else if (type.name == "Float") {
            return if (value is RecordValue.Number) {
                true to value.value
            } else {
                false to null
            }
        } else if (type.name == "ID") {
            return if (value is RecordValue.String) {
                true to value.value
            } else {
                false to null
            }
        } else if (type.name == "Date") {
            return if (value is RecordValue.String) {
                true to value.value
            } else {
                false to null
            }
        } else if (type.name == "Boolean") {
            return if (value is RecordValue.Boolean) {
                true to value.value
            } else {
                false to null
            }
        } else {
            error("Unknown scalar type: " + type.name)
        }
    } else if (type is OutputType.NotNull) {
        return if (value == RecordValue.Null) {
            false to null
        } else {
            readValue(value, type.inner, store, arguments)
        }
    } else if (type is OutputType.List) {
        if (value == RecordValue.Null) {
            return true to null
        } else {
            if (value is RecordValue.List) {
                val mapped = value.items.map { readValue(it, type.inner, store, arguments) }
                if (mapped.any { !it.first }) {
                    return false to null
                }
                val v = mapped.map { it.second!! }
                return true to JSONArray(v)
            } else {
                error("Invalid record value")
            }
        }
    } else if (type is OutputType.Object) {
        if (value == RecordValue.Null) {
            return true to null
        } else {
            if (value is RecordValue.Reference) {
                return readSelector(value.key, store, type.selectors, arguments)
            } else {
                error("Invalid record value")
            }
        }
    }
    return false to null
}

private fun readSelector(
        record: Record,
        fields: MutableMap<String, Any?>,
        store: RecordStore,
        selectors: List<Selector>,
        arguments: JSONObject
): Boolean {
    for (f in selectors) {
        if (f is Selector.Field) {
            val key = selectorKey(f.name, f.arguments, arguments)
            if (record.fields.containsKey(key)) {
                val rv = readValue(record.fields[key]!!, f.type, store, arguments)
                if (!rv.first) {
                    return false
                }
                fields[f.alias] = rv.second
            } else {
                return false
            }
        } else if (f is Selector.TypeCondition) {
            if (record.fields["__typename"] == RecordValue.String(f.type)) {
                if (!readSelector(record, fields, store, f.fragment.selectors, arguments)) {
                    return false
                }
            }
        } else if (f is Selector.Fragment) {
            // Can't check for type since there typename could be subclass
//            if (record.fields["__typename"] != RecordValue.String(f.type)) {
//                return false
//            }
            if (!readSelector(record, fields, store, f.fragment.selectors, arguments)) {
                return false
            }
        } else {
            error("Unsupported selector")
        }
    }
    return true
}

private fun readSelector(cacheKey: String, store: RecordStore, selectors: List<Selector>, arguments: JSONObject): Pair<Boolean, JSONObject?> {
    val value = store.read(cacheKey)
    if (value.fields.isEmpty()) {
        return false to null
    }
    val k = store.read(value.key)
    val fields = mutableMapOf<String, Any?>()
    if (!readSelector(k, fields, store, selectors, arguments)) {
        return false to null
    }
    return true to JSONObject(fields)
}

fun readFromStore(cacheKey: String, store: RecordStore, type: OutputType.Object, arguments: JSONObject): Pair<Boolean, JSONObject?> {
    return readSelector(cacheKey, store, type.selectors, arguments)
}

File: ios/readFromStore.swift
//
//  RecordStoreReader.swift
//  openland
//
//  Created by Steve Kite on 5/13/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON

enum StoreReadResult {
  case success(data: JSON)
  case missing
}

fileprivate func readValue(
  value: RecordValue,
  type: OutputType,
  store: RecordStore,
  variables: JSON
) -> StoreReadResult {
  if type is OutputType.Scalar {
    let sc = type as! OutputType.Scalar
    if value is RecordValue.NullValue {
      return StoreReadResult.success(data: JSON(NSNull()))
    } else if sc.name == "String" {
      if value is RecordValue.StringValue {
        return StoreReadResult.success(data: JSON((value as! RecordValue.StringValue).value))
      } else {
        return StoreReadResult.missing
      }
    } else if sc.name == "Int" {
      if value is RecordValue.NumberValue {
        return StoreReadResult.success(data: JSON((value as! RecordValue.NumberValue).value))
      } else {
        return StoreReadResult.missing
      }
    } else if sc.name == "Float" {
      if value is RecordValue.NumberValue {
        return StoreReadResult.success(data: JSON((value as! RecordValue.NumberValue).value))
      } else {
        return StoreReadResult.missing
      }
    } else if sc.name == "ID" {
      if value is RecordValue.StringValue {
        return StoreReadResult.success(data: JSON((value as! RecordValue.StringValue).value))
      } else {
        return StoreReadResult.missing
      }
    } else if sc.name == "Date" {
      if value is RecordValue.StringValue {
        return StoreReadResult.success(data: JSON((value as! RecordValue.StringValue).value))
      } else {
        return StoreReadResult.missing
      }
    } else if sc.name == "Boolean" {
      if value is RecordValue.BooleanValue {
        return StoreReadResult.success(data: JSON((value as! RecordValue.BooleanValue).value))
      } else {
        return StoreReadResult.missing
      }
    } else {
      fatalError("Unknown scalar type: " + sc.name)
    }
  } else if type is OutputType.NotNull {
    if value is RecordValue.NullValue {
      return StoreReadResult.missing
    } else {
      return readValue(
        value: value,
        type: (type as! OutputType.NotNull).inner,
        store: store,
        variables: variables
      )
    }
  } else if type is OutputType.List {
    if value is RecordValue.NullValue {
      return StoreReadResult.success(data: JSON(NSNull()))
    } else {
      if value is RecordValue.ListValue {
        let lt = type as! OutputType.List
        let l = value as! RecordValue.ListValue
        var res: [JSON] = []
        for i in l.items {
          let r = readValue(value: i, type: lt.inner, store: store, variables: variables)
          switch(r) {
          case .success(let data):
            res.append(data)
          case .missing:
            return StoreReadResult.missing
          }
        }
        return StoreReadResult.success(data: JSON(res))
      } else {
        return StoreReadResult.missing
      }
    }
  } else if type is OutputType.Object {
    if value is RecordValue.NullValue {
      return StoreReadResult.success(data: JSON(NSNull()))
    } else {
      let ot = type as! OutputType.Object
      if value is RecordValue.ReferenceValue {
        return readSelector(cacheKey: (value as! RecordValue.ReferenceValue).key, store: store, selectors: ot.selectors, variables: variables)
      } else {
        return StoreReadResult.missing
      }
    }
  }
  
  fatalError()
}

fileprivate func readSelector(
  record: Record,
  fields: SharedDictionary<String, JSON>,
  store: RecordStore,
  selectors: [Selector],
  variables: JSON
) -> Bool {
  for f in selectors {
    if f is Selector.Field {
      let fl = f as! Selector.Field
      let key = selectorKey(name: fl.name, arguments: fl.arguments, variables: variables)
      if record.fields[key] != nil {
        let rv = readValue(value: record.fields[key]!, type: fl.type, store: store, variables: variables)
        switch(rv) {
        case .success(let data):
          fields[fl.alias] = data
        case .missing:
          return false
        }
      } else {
        return false
      }
    } else if f is Selector.TypeCondition {
      let tc = f as! Selector.TypeCondition
      if record.fields["__typename"] == RecordValue.StringValue(value: tc.name) {
        if !readSelector(record: record, fields: fields, store: store, selectors: tc.type.selectors, variables: variables) {
          return false
        }
      }
    } else if f is Selector.Fragment {
      let ff = f as! Selector.Fragment
      if !readSelector(record: record, fields: fields, store: store, selectors: ff.type.selectors, variables: variables) {
        return false
      }
    } else {
      fatalError("Unsupported selector")
    }
  }
  return true
}

fileprivate func readSelector(
  cacheKey: String,
  store:RecordStore,
  selectors: [Selector],
  variables: JSON
  ) -> StoreReadResult {

  let value = store.read(key: cacheKey)
  if value.fields.count == 0{
    return StoreReadResult.missing
  }
  
  let fields = SharedDictionary<String, JSON>()
  
  if !readSelector(record: value, fields: fields, store: store, selectors: selectors, variables: variables) {
    return StoreReadResult.missing
  }
  
  return StoreReadResult.success(data: JSON(fields.dict))
}

func readFromStore(cacheKey: String, store: RecordStore, type: OutputType.Object, variables: JSON) -> StoreReadResult {
  return readSelector(cacheKey: cacheKey, store: store, selectors: type.selectors, variables: variables)
}

func readQueryFromStore(cacheKey: String, store: RecordStore, type: OutputType.Object, variables: JSON) -> StoreReadResult {
  var fields: [String: JSON] = [:]
  for f in type.selectors {
    if !(f is Selector.Field) {
      fatalError("Root query can't contain fragments")
    }
    let sf = f as! Selector.Field
    let key = selectorKey(name: sf.name, arguments: sf.arguments, variables: variables)
    let refId = "\(cacheKey).$ref.\(key)"
    let value = store.read(key: refId)
    let ex = value.fields["data"]
    if ex == nil {
      // NSLog("[SpaceX-Store]: Missing query \(refId)")
      return StoreReadResult.missing
    }
    let rv = readValue(value: ex!, type: sf.type, store: store, variables: variables)
    switch(rv) {
    case .missing:
      // NSLog("[SpaceX-Store]: Missing value")
      return StoreReadResult.missing
    case .success(let data):
      fields[sf.alias] = data
    }
  }
  return StoreReadResult.success(data: JSON(fields))
}
